File: keprix_sdk/typescript/src/sidecar.ts
/** Keprix Universal Sidecar HTTP client (`/sidecar/v1`). */

export type SidecarClientOptions = {
  baseUrl: string;
  token?: string;
  projectKey?: string;
  timeoutMs?: number;
  webhookSecret?: string;
};

export class SidecarClient {
  baseUrl: string;
  token: string;
  projectKey?: string;
  timeoutMs: number;
  webhookSecret?: string;
  private prefix: string;

  constructor(options: SidecarClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.token = options.token || "";
    this.projectKey = options.projectKey;
    this.timeoutMs = options.timeoutMs ?? 60_000;
    this.webhookSecret = options.webhookSecret;
    this.prefix = `${this.baseUrl}/sidecar/v1`;
  }

  private headers(extra?: Record<string, string>): Record<string, string> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.token) headers.Authorization = `Bearer ${this.token}`;
    return { ...headers, ...(extra || {}) };
  }

  private project(projectKey?: string): string {
    const key = projectKey || this.projectKey;
    if (!key) throw new Error("projectKey is required");
    return key;
  }

  private async request(
    method: string,
    path: string,
    body?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(`${this.prefix}${path}`, {
        method,
        headers: this.headers(),
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: controller.signal,
      });
      const text = await response.text();
      if (!response.ok) {
        throw new Error(`${method} ${path} failed: ${response.status} ${text}`);
      }
      if (!text) return {};
      return JSON.parse(text) as Record<string, unknown>;
    } finally {
      clearTimeout(timer);
    }
  }

  async pairBootstrap(params: {
    pairingCode: string;
    projectKey?: string;
    deployment?: string;
    environment?: string;
    requestedScopes?: string[];
  }): Promise<Record<string, unknown>> {
    const result = await this.request("POST", "/pair/bootstrap", {
      pairing_code: params.pairingCode,
      project_key: this.project(params.projectKey),
      deployment: params.deployment || "local-dev",
      environment: params.environment || "local",
      requested_scopes: params.requestedScopes,
    });
    const token = (result.access_token || result.token) as string | undefined;
    if (token) this.token = token;
    return result;
  }

  async health(projectKey?: string): Promise<Record<string, unknown>> {
    const key = projectKey || this.projectKey;
    if (key) return this.request("GET", `/projects/${key}/health`);
    return this.request("GET", "/health");
  }

  async capabilities(projectKey?: string): Promise<Record<string, unknown>> {
    const key = this.project(projectKey);
    return this.request("GET", `/projects/${key}/capabilities`);
  }

  async session(params: {
    purpose: string;
    tenantId: string;
    actorId: string;
    projectKey?: string;
    grants?: string[];
  }): Promise<Record<string, unknown>> {
    const key = this.project(params.projectKey);
    return this.request("POST", `/projects/${key}/sessions`, {
      purpose: params.purpose,
      tenant_id: params.tenantId,
      actor_id: params.actorId,
      grants: params.grants,
    });
  }

  async invoke(params: {
    node: string;
    input?: Record<string, unknown>;
    purpose?: string;
    projectKey?: string;
    sessionId?: string;
  }): Promise<Record<string, unknown>> {
    const key = this.project(params.projectKey);
    return this.request("POST", `/projects/${key}/invoke`, {
      node: params.node,
      input: params.input || {},
      purpose: params.purpose || "invoke",
      session_id: params.sessionId,
    });
  }

  async jobs(params: {
    node: string;
    input?: Record<string, unknown>;
    purpose?: string;
    projectKey?: string;
    idempotencyKey?: string;
  }): Promise<Record<string, unknown>> {
    const key = this.project(params.projectKey);
    return this.request("POST", `/projects/${key}/jobs`, {
      node: params.node,
      input: params.input || {},
      purpose: params.purpose || "job",
      idempotency_key: params.idempotencyKey,
    });
  }

  async getJob(jobId: string, projectKey?: string): Promise<Record<string, unknown>> {
    const key = this.project(projectKey);
    return this.request("GET", `/projects/${key}/jobs/${jobId}`);
  }

  async cancel(jobId: string, projectKey?: string): Promise<Record<string, unknown>> {
    const key = this.project(projectKey);
    return this.request("POST", `/projects/${key}/jobs/${jobId}/cancel`, {});
  }

  async sendEvent(
    event: Record<string, unknown>,
    projectKey?: string,
  ): Promise<Record<string, unknown>> {
    const key = this.project(projectKey);
    return this.request("POST", `/projects/${key}/events`, event);
  }

  async *streamEvents(projectKey?: string): AsyncGenerator<Record<string, unknown>> {
    const key = this.project(projectKey);
    const response = await fetch(`${this.prefix}/projects/${key}/events/stream`, {
      headers: this.headers(),
    });
    if (!response.ok || !response.body) {
      throw new Error(`stream failed: ${response.status}`);
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    let dataLines: string[] = [];
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() || "";
      for (const line of lines) {
        if (line.startsWith("data:")) {
          dataLines.push(line.slice(5).trimStart());
        } else if (line === "" && dataLines.length) {
          const raw = dataLines.join("\n");
          dataLines = [];
          try {
            yield JSON.parse(raw) as Record<string, unknown>;
          } catch {
            yield { raw };
          }
        }
      }
    }
  }

  async approvalDecision(params: {
    approvalId: string;
    decision: string;
    actorId: string;
    projectKey?: string;
    reason?: string;
  }): Promise<Record<string, unknown>> {
    const key = this.project(params.projectKey);
    return this.request("POST", `/projects/${key}/approvals/${params.approvalId}/decision`, {
      decision: params.decision,
      actor_id: params.actorId,
      reason: params.reason,
    });
  }

  async verifyWebhook(params: {
    body: string | ArrayBuffer | Uint8Array;
    signatureHeader: string;
    timestampHeader?: string;
    maxSkewSeconds?: number;
    secret?: string;
  }): Promise<boolean> {
    const secret = params.secret || this.webhookSecret;
    if (!secret) throw new Error("webhook secret required");
    const enc = new TextEncoder();
    let raw: Uint8Array;
    if (typeof params.body === "string") raw = enc.encode(params.body);
    else if (params.body instanceof ArrayBuffer) raw = new Uint8Array(params.body);
    else raw = params.body;

    let provided = params.signatureHeader.trim();
    let ts: string | undefined;
    if (provided.includes("v1=")) {
      const parts: Record<string, string> = {};
      for (const piece of provided.split(",")) {
        const [k, v] = piece.split("=", 2);
        if (k && v) parts[k] = v;
      }
      ts = parts.t;
      provided = parts.v1 || "";
    }
    if (params.timestampHeader && ts === undefined) ts = params.timestampHeader;
    const maxSkew = params.maxSkewSeconds ?? 300;
    let signed: Uint8Array = raw;
    if (ts !== undefined) {
      const tsNum = Number(ts);
      if (!Number.isFinite(tsNum) || Math.abs(Date.now() / 1000 - tsNum) > maxSkew) {
        return false;
      }
      const prefix = enc.encode(`${ts}.`);
      signed = new Uint8Array(prefix.length + raw.length);
      signed.set(prefix, 0);
      signed.set(raw, prefix.length);
    }
    const key = await crypto.subtle.importKey(
      "raw",
      enc.encode(secret),
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["sign"],
    );
    const sig = await crypto.subtle.sign("HMAC", key, signed);
    const hex = Array.from(new Uint8Array(sig))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
    return hex === provided.toLowerCase();
  }

  async connectorTest(params: {
    connectorKey: string;
    pathParams?: Record<string, unknown>;
    projectKey?: string;
  }): Promise<Record<string, unknown>> {
    const key = this.project(params.projectKey);
    return this.request("POST", `/projects/${key}/connectors/${params.connectorKey}/test`, {
      path_params: params.pathParams || {},
    });
  }
}

File: keprix_sdk/python/src/keprix_sdk/sidecar.py
"""Keprix Universal Sidecar HTTP client."""

from __future__ import annotations

import hashlib
import hmac
import json
import time
from collections.abc import Iterator
from typing import Any
from urllib.parse import urljoin

import httpx


class SidecarClient:
    """Client for `/sidecar/v1` (mounted :3333 or sidecar-only :3360)."""

    def __init__(
        self,
        base_url: str,
        token: str | None = None,
        *,
        project_key: str | None = None,
        timeout: float = 60.0,
        webhook_secret: str | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.token = token or ""
        self.project_key = project_key
        self.timeout = timeout
        self.webhook_secret = webhook_secret
        self._prefix = f"{self.base_url}/sidecar/v1"

    def _headers(self, extra: dict[str, str] | None = None) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        if extra:
            headers.update(extra)
        return headers

    def _project(self, project_key: str | None = None) -> str:
        key = project_key or self.project_key
        if not key:
            raise ValueError("project_key is required")
        return key

    def _request(
        self,
        method: str,
        path: str,
        *,
        json_body: dict[str, Any] | None = None,
        stream: bool = False,
    ) -> Any:
        url = urljoin(self._prefix + "/", path.lstrip("/"))
        with httpx.Client(timeout=self.timeout) as client:
            response = client.request(
                method,
                url,
                headers=self._headers(),
                json=json_body,
            )
            if stream:
                response.raise_for_status()
                return response
            if response.status_code >= 400:
                raise RuntimeError(f"{method} {path} failed: {response.status_code} {response.text}")
            if not response.content:
                return {}
            return response.json()

    def pair_bootstrap(
        self,
        pairing_code: str,
        *,
        project_key: str | None = None,
        deployment: str = "local-dev",
        environment: str = "local",
        requested_scopes: list[str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "pairing_code": pairing_code,
            "project_key": self._project(project_key),
            "deployment": deployment,
            "environment": environment,
        }
        if requested_scopes is not None:
            body["requested_scopes"] = requested_scopes
        result = self._request("POST", "/pair/bootstrap", json_body=body)
        token = result.get("access_token") or result.get("token")
        if isinstance(token, str) and token:
            self.token = token
        return result

    def health(self, project_key: str | None = None) -> dict[str, Any]:
        key = project_key or self.project_key
        if key:
            return self._request("GET", f"/projects/{key}/health")
        return self._request("GET", "/health")

    def capabilities(self, project_key: str | None = None) -> dict[str, Any]:
        key = self._project(project_key)
        return self._request("GET", f"/projects/{key}/capabilities")

    def session(
        self,
        *,
        purpose: str,
        tenant_id: str,
        actor_id: str,
        project_key: str | None = None,
        grants: list[str] | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        body: dict[str, Any] = {
            "purpose": purpose,
            "tenant_id": tenant_id,
            "actor_id": actor_id,
        }
        if grants is not None:
            body["grants"] = grants
        return self._request("POST", f"/projects/{key}/sessions", json_body=body)

    def invoke(
        self,
        node: str,
        input: dict[str, Any] | None = None,
        *,
        purpose: str = "invoke",
        project_key: str | None = None,
        session_id: str | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        body: dict[str, Any] = {"node": node, "input": input or {}, "purpose": purpose}
        if session_id:
            body["session_id"] = session_id
        return self._request("POST", f"/projects/{key}/invoke", json_body=body)

    def jobs(
        self,
        node: str,
        input: dict[str, Any] | None = None,
        *,
        purpose: str = "job",
        project_key: str | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        body: dict[str, Any] = {"node": node, "input": input or {}, "purpose": purpose}
        if idempotency_key:
            body["idempotency_key"] = idempotency_key
        return self._request("POST", f"/projects/{key}/jobs", json_body=body)

    def get_job(self, job_id: str, *, project_key: str | None = None) -> dict[str, Any]:
        key = self._project(project_key)
        return self._request("GET", f"/projects/{key}/jobs/{job_id}")

    def cancel(self, job_id: str, *, project_key: str | None = None) -> dict[str, Any]:
        key = self._project(project_key)
        return self._request("POST", f"/projects/{key}/jobs/{job_id}/cancel", json_body={})

    def send_event(
        self,
        event: dict[str, Any],
        *,
        project_key: str | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        return self._request("POST", f"/projects/{key}/events", json_body=event)

    def stream_events(
        self,
        *,
        project_key: str | None = None,
    ) -> Iterator[dict[str, Any]]:
        """Iterate SSE JSON payloads from `/events/stream`."""
        key = self._project(project_key)
        url = f"{self._prefix}/projects/{key}/events/stream"
        with httpx.Client(timeout=None) as client:
            with client.stream("GET", url, headers=self._headers()) as response:
                response.raise_for_status()
                data_lines: list[str] = []
                for line in response.iter_lines():
                    if line is None:
                        continue
                    if line.startswith("data:"):
                        data_lines.append(line[5:].lstrip())
                    elif line == "" and data_lines:
                        raw = "\n".join(data_lines)
                        data_lines = []
                        try:
                            yield json.loads(raw)
                        except json.JSONDecodeError:
                            yield {"raw": raw}

    def approval_decision(
        self,
        approval_id: str,
        decision: str,
        *,
        actor_id: str,
        project_key: str | None = None,
        reason: str | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        body: dict[str, Any] = {"decision": decision, "actor_id": actor_id}
        if reason:
            body["reason"] = reason
        return self._request(
            "POST",
            f"/projects/{key}/approvals/{approval_id}/decision",
            json_body=body,
        )

    def verify_webhook(
        self,
        body: bytes | str,
        signature_header: str,
        *,
        timestamp_header: str | None = None,
        max_skew_seconds: int = 300,
        secret: str | None = None,
    ) -> bool:
        """Verify HMAC-SHA256 webhook signature (`t=...,v1=...` or raw hex)."""
        secret_value = secret or self.webhook_secret
        if not secret_value:
            raise ValueError("webhook secret required")
        raw = body.encode("utf-8") if isinstance(body, str) else body
        ts = None
        provided = signature_header.strip()
        if "v1=" in provided:
            parts = dict(
                piece.split("=", 1) for piece in provided.split(",") if "=" in piece
            )
            ts = parts.get("t")
            provided = parts.get("v1", "")
        if timestamp_header and ts is None:
            ts = timestamp_header
        if ts is not None:
            try:
                if abs(int(time.time()) - int(ts)) > max_skew_seconds:
                    return False
            except ValueError:
                return False
            signed = f"{ts}.".encode("utf-8") + raw
        else:
            signed = raw
        digest = hmac.new(secret_value.encode("utf-8"), signed, hashlib.sha256).hexdigest()
        return hmac.compare_digest(digest, provided.lower())

    def connector_test(
        self,
        connector_key: str,
        *,
        path_params: dict[str, Any] | None = None,
        project_key: str | None = None,
    ) -> dict[str, Any]:
        key = self._project(project_key)
        return self._request(
            "POST",
            f"/projects/{key}/connectors/{connector_key}/test",
            json_body={"path_params": path_params or {}},
        )
